File: branches/convenience-views/implementation/UrakawaToolkit/urakawa/media/timing/ITime.cs
using System;

namespace urakawa.media.timing
{
	/// <summary>
	/// A simple interface for a measure of time.
	/// </summary>
	public interface ITime
	{
		/// <summary>
		/// Determines if the time offset is negative.
		/// </summary>
		/// <returns></returns>
		bool isNegativeTimeOffset();

		/// <summary>
		/// Copy this object.
		/// </summary>
		/// <returns></returns>
		ITime copy();

		/// <summary>
		/// Gets the (signed) <see cref="ITimeDelta"/> between a given <see cref="ITime"/> and <c>this</c>,
		/// that is <c>this-<paramref localName="t"/></c>
		/// </summary>
		/// <param localName="t">The given <see cref="ITime"/></param>
		/// <returns>
		/// The difference as an <see cref="ITimeDelta"/>
		/// </returns>
		ITimeDelta getTimeDelta(ITime t);

		/// <summary>
		/// Gets number of milliseconds equivalent of the <see cref="ITime"/>
		/// </summary>
		/// <returns>The number of milliseconds</returns>
		long getTimeAsMilliseconds();

		/// <summary>
		/// Gets the millisecond floating point number equivalent of the <see cref="ITime"/>
		/// </summary>
		/// <returns>The millisecond floating point number</returns>
		double getTimeAsMillisecondFloat();

		/// <summary>
		/// Sets the <see cref="ITime"/> from an integral number of milliseconds
		/// </summary>
		/// <param localName="timeAsMS">The number of milliseconds</param>
		void setTime(long timeAsMS);

		/// <summary>
		/// Sets the <see cref="ITime"/> from a floating point mulliseconds value
		/// </summary>
		/// <param localName="timeAsMSF">The milliseconds floating point number</param>
		void setTime(double timeAsMSF);

		/// <summary>
		/// Determines is <c>this</c> is greater than a given other <see cref="ITime"/>.
		/// </summary>
		/// <param localName="otherTime">The other <see cref="ITime"/></param>
		/// <returns>
		/// <c>true</c> if <c>this</c> is greater than <paramref localName="otherTime"/>, otherwise <c>false</c>
		/// </returns>
		bool isGreaterThan(ITime otherTime);

		/// <summary>
		/// Determines is <c>this</c> is less than a given other <see cref="ITime"/>.
		/// </summary>
		/// <param localName="otherTime">The other <see cref="ITime"/></param>
		/// <returns>
		/// <c>true</c> if <c>this</c> is less than <paramref localName="otherTime"/>, otherwise <c>false</c>
		/// </returns>
		/// <remarks>
		/// <c>isLessThan(t) = t.isGreaterThan(this)</c> for any <see cref="ITime"/> <c>t</c>
		/// </remarks>
		bool isLessThan(ITime otherTime);

		/// <summary>
		/// Determines is <c>this</c> value equal to a given other <see cref="ITime"/>
		/// </summary>
		/// <param localName="otherTime">The other <see cref="ITime"/></param>
		/// <returns>
		/// <c>true</c> if <c>this</c> and <paramref localName="otherTime"/> are value equal,
		/// otherwise <c>false</c>
		/// </returns>
		/// <remarks>
		/// <c>isEqualTo(t) = !(isGreaterThan(t) || t.isGreaterThan(this))</c>
		/// for any <see cref="ITime"/> <c>t</c>
		/// </remarks>
		bool isEqualTo(ITime otherTime);
	}
}

File: java/core/src/org/daisy/urakawa/media/timing/ITime.java
package org.daisy.urakawa.media.timing;

import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * ITime offset (could be in milliseconds, SMPTE, etc.). This really is an
 * interface "lollypop" that should be extended. Typically, methods like
 * getTimeMilliseconds(), getTimeSMPTE(), etc. should be available to the
 * end-user of the API. Can be a negative/0/positive offset relative to the
 * local timebase in the current context.
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Clone - org.daisy.urakawa.media.timing.ITime
 */
public interface ITime
{
    /**
     * @return the "zero" time
     */
    public ITime getZero();

    /**
     * @return the maximum time value
     */
    public ITime getMaxValue();

    /**
     * @return the minimum time value
     */
    public ITime getMinValue();

    /**
     * @return time
     */
    public long getTimeAsMilliseconds();

    /**
     * @param otherTime
     * @return true or false
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public boolean isLessThanOrEqualTo(ITime otherTime)
            throws MethodParameterIsNullException;

    /**
     * @param otherTime
     * @return true or false
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public boolean isGreaterThanOrEqualTo(ITime otherTime)
            throws MethodParameterIsNullException;

    /**
     * @param other
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public ITime subtractTime(ITime other)
            throws MethodParameterIsNullException;

    /**
     * @param other
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    public ITime subtractTimeDelta(ITimeDelta other)
            throws MethodParameterIsNullException;

    /**
     * @param newTime
     */
    public void setTime(long newTime);

    /**
     * @param stringRepresentation
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     * @throws MethodParameterIsEmptyStringException
     * @throws TimeStringRepresentationIsInvalidException
     */
    public ITime parse(String stringRepresentation)
            throws MethodParameterIsNullException,
            MethodParameterIsEmptyStringException,
            TimeStringRepresentationIsInvalidException;

    /**
     * a helper method to help determine
     * {@link org.daisy.urakawa.media.timing.TimeOffsetIsNegativeException}
     * 
     * @return true if the associated time value is a negative offset (<0 "less
     *         than zero")
     */
    public boolean isNegativeTimeOffset();

    /**
     * <p>
     * Cloning method
     * </p>
     * 
     * @return a copy.
     */
    ITime copy();

    /**
     * @param t
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    ITimeDelta getTimeDelta(ITime t) throws MethodParameterIsNullException;

    /**
     * @param other
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    ITime addTime(ITime other) throws MethodParameterIsNullException;

    /**
     * @param other
     * @return time
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    ITime addTimeDelta(ITimeDelta other) throws MethodParameterIsNullException;

    /**
     * @param otherTime
     * @return true or false
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    boolean isGreaterThan(ITime otherTime)
            throws MethodParameterIsNullException;

    /**
     * @param otherTime
     * @return true or false
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    boolean isLessThan(ITime otherTime) throws MethodParameterIsNullException;

    /**
     * @param otherTime
     * @return true or false
     * @throws MethodParameterIsNullException
     *         NULL method parameters are forbidden
     */
    boolean isEqualTo(ITime otherTime) throws MethodParameterIsNullException;
}